src/header/common/content_disposition.rs

```diff
@@ -65,88 +218,215 @@ pub enum DispositionParam {
-            let mut sections = s.split(';');
-            let disposition = match sections.next() {
-                Some(s) => s.trim(),
-                None => return Err(::error::ParseError::Header),
-            };
-
-            let mut cd = ContentDisposition {
-                disposition: if disposition.eq_ignore_ascii_case("inline") {
-                    DispositionType::Inline
-                } else if disposition.eq_ignore_ascii_case("attachment") {
-                    DispositionType::Attachment
-                } else {
-                    DispositionType::Ext(disposition.to_owned())
-                },
-                parameters: Vec::new(),
-            };
-
-            for section in sections {
-                let mut parts = section.splitn(2, '=');
+        // `header::from_one_raw_str` invokes `hv.to_str` which assumes `hv` contains only visible
+        //  ASCII characters. So `hv.as_bytes` is necessary here.
+        let hv = String::from_utf8(hv.as_bytes().to_vec())
+            .map_err(|_| ::error::ParseError::Header)?;
+        let (disp_type, mut left) = split_once_and_trim(hv.as_str().trim(), ';');
+        if disp_type.len() == 0 {
+            return Err(::error::ParseError::Header);
+        }
+        let mut cd = ContentDisposition {
+            disposition: disp_type.into(),
+            parameters: Vec::new(),
+        };
 
-                let key = if let Some(key) = parts.next() {
-                    key.trim()
+        while left.len() > 0 {
+            let (param_name, new_left) = split_once_and_trim(left, '=');
+            if param_name.len() == 0 {
+                return Err(::error::ParseError::Header);
+            }
+            left = new_left;
+            if param_name.ends_with('*') {
+                // extended parameters
+                let (param_name, _) = param_name.split_at(param_name.len() - 1); // trim asterisk
+                let (ext_value, new_left) = split_once_and_trim(left, ';');
+                left = new_left;
+                let ext_value = header::parse_extended_value(ext_value)?;
+                cd.parameters
+                    .push(match param_name.to_ascii_lowercase().as_str() {
+                        "filename" => DispositionParam::FilenameExt(ext_value),
+                        _ => DispositionParam::UnknownExt(
+                            param_name.to_owned(),
+                            ext_value,
+                        ),
+                    });
+            } else {
+                // regular parameters
+                let value = if left.starts_with('\"') {
+                    // quoted-string: defined in RFC6266 -> RFC2616 Section 3.6
+                    let mut escaping = false;
+                    let mut quoted_string = vec![];
+                    let mut end = None;
+                    // search for closing quote
+                    for (i, &c) in left.as_bytes().iter().skip(1).enumerate() {
```

Comment:
why use reference for `&c`?

Reply:
To dereference `&u8` into `u8`? I think `as_bytes().iter()` produces &u8 instead of u8, am I right?
I am not sure how to write that idiomatically in Rust.

Reply:
After change `&c` to `c` in line 311:
```
error[E0277]: can't compare `&u8` with `{integer}`
   --> src/header/common/content_disposition.rs:316:34
    |
316 |                             if c == 0x5c // backslash
    |                                  ^^ no implementation for `&u8 == {integer}`
    |
    = help: the trait `std::cmp::PartialEq<{integer}>` is not implemented for `&u8`

error[E0277]: can't compare `&u8` with `{integer}`
   --> src/header/common/content_disposition.rs:319:41
    |
319 |                             } else if c == 0x22 // double quote
    |                                         ^^ no implementation for `&u8 == {integer}`
    |
    = help: the trait `std::cmp::PartialEq<{integer}>` is not implemented for `&u8`

error[E0308]: mismatched types
   --> src/header/common/content_disposition.rs:331:59
    |
331 |                     let quoted_string = String::from_utf8(quoted_string)
    |                                                           ^^^^^^^^^^^^^ expected u8, found &u8
    |
    = note: expected type `std::vec::Vec<u8>`
               found type `std::vec::Vec<&u8>`

```

Reply:
Aha, I am often messed up by those rules/implications too.